src/api/extension/wallet.js

```diff
@@ -496,3 +496,155 @@ export const withdrawalTx = async (account, delegation, protocolParameters) => {
 };
+
+export const undelegateTx = async (account, delegation, protocolParameters) => {
+  await Loader.load();
+  const utxos = await getUtxos();
+  //estimated max multiasset size 5848
+  //estimated max value size 5860
+  //estimated max utxo size 5980
+  const MULTIASSET_SIZE = 5848;
+  const VALUE_SIZE = 5860;
+
+  const outputs = Loader.Cardano.TransactionOutputs.new();
+  outputs.add(
+    Loader.Cardano.TransactionOutput.new(
+      Loader.Cardano.Address.from_bech32(account.paymentAddr),
+      Loader.Cardano.Value.new(
+        Loader.Cardano.BigNum.from_str(protocolParameters.minUtxo)
+      )
+    )
+  );
+  CoinSelection.setProtocolParameters(
+    protocolParameters.coinsPerUtxoWord,
+    protocolParameters.linearFee.minFeeA,
+    protocolParameters.linearFee.minFeeB,
+    protocolParameters.maxTxSize.toString()
+  );
+  const selection = await CoinSelection.randomImprove(utxos, outputs, 20);
+
+  const inputs = selection.input;
+  const txBuilder = Loader.Cardano.TransactionBuilder.new(
+    Loader.Cardano.LinearFee.new(
+      Loader.Cardano.BigNum.from_str(protocolParameters.linearFee.minFeeA),
+      Loader.Cardano.BigNum.from_str(protocolParameters.linearFee.minFeeB)
+    ),
+    Loader.Cardano.BigNum.from_str(protocolParameters.poolDeposit),
+    Loader.Cardano.BigNum.from_str(protocolParameters.keyDeposit),
+    protocolParameters.maxValSize,
+    protocolParameters.maxTxSize,
+    Loader.Cardano.BigNum.from_str(protocolParameters.coinsPerUtxoWord)
+  );
+  for (let i = 0; i < inputs.length; i++) {
+    const utxo = inputs[i];
+    txBuilder.add_input(
+      utxo.output().address(),
+      utxo.input(),
+      utxo.output().amount()
+    );
+  }
+
+  if (delegation.rewards > 0) {
+    const withdrawals = Loader.Cardano.Withdrawals.new();
+    withdrawals.insert(
+      Loader.Cardano.RewardAddress.from_address(
+        Loader.Cardano.Address.from_bech32(account.rewardAddr)
+      ),
+      Loader.Cardano.BigNum.from_str(delegation.rewards)
+    );
+
+    txBuilder.set_withdrawals(withdrawals);
+  }
+
+  const certificates = Loader.Cardano.Certificates.new();
+
+  certificates.add(
+    Loader.Cardano.Certificate.new_stake_deregistration(
+      Loader.Cardano.StakeDeregistration.new(
+        Loader.Cardano.StakeCredential.from_keyhash(
+          Loader.Cardano.Ed25519KeyHash.from_bytes(
+            Buffer.from(account.stakeKeyHash, 'hex')
+          )
+        )
+      )
+    )
+  );
+
+  txBuilder.set_certs(certificates);
+
+  const change = selection.change;
+  const changeMultiAssets = change.multiasset();
+
+  // check if change value is too big for single output
+  if (
+    changeMultiAssets &&
+    change.to_bytes().length * 2 > protocolParameters.maxValSize
+  ) {
+    const partialChange = Loader.Cardano.Value.new(
+      Loader.Cardano.BigNum.from_str('0')
+    );
+
+    const partialMultiAssets = Loader.Cardano.MultiAsset.new();
+    const policies = changeMultiAssets.keys();
+    const makeSplit = () => {
```

Comment:
Would it make sense to move this function outside `undelegateTx’? It seems like something that could be used in the future. It also could be unit-tested independently.

Reply:
Hi! Thanks for reviewing the code. It will benefit of a large refactor indeed, let say all of this was pretty experimental at first.
The makeSplit function is actually removed on the cs-sl-split branch.
If we do move forward with the branch, this part will be removed.

Reply:
oh ok ok! — I hope if you don’t mind if I appear in this repo more and more. I want to contribute to Cardano projects and working with JS repos is something familiar to me.

Reply:
Alright, feel free to provide help as you like.
The project is actually opened source and there is a lot to do and improve, so help is very much appreciated.
Alessandro is the lead maintainer, so before initiating any work make sure to checkout with him to get the pulse.

Reply:
Thank you!